master/internal/db/cache.go

```diff
@@ -0,0 +1,187 @@
+package db
+
+import (
+	"context"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/proto/pkg/experimentv1"
+)
+
+type modelDefFolder struct {
+	fileTree   []*experimentv1.FileNode
+	cachedTime time.Time
+	lock       sync.RWMutex
+	path       string
+}
+
+// FileCache is metadata for files cached at file system.
+type FileCache struct {
+	rootDir string
+	maxAge  time.Duration
+	caches  map[int]*modelDefFolder
+	lock    sync.Mutex
+}
+
+// NewFileCache initialize FileCache obj.
+func NewFileCache(rootDir string, maxAge time.Duration) *FileCache {
+	return &FileCache{
+		rootDir: rootDir,
+		maxAge:  maxAge,
+		caches:  make(map[int]*modelDefFolder),
+	}
+}
+
+func (f *FileCache) getOrCreateFolder(expID int) (*modelDefFolder, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	value, ok := f.caches[expID]
+	if !ok {
+		exp := struct {
+			ModelDefinition []byte
+		}{}
+		err := Bun().NewSelect().TableExpr(
+			"experiments").Column("model_definition").Where("id = ?", expID).Scan(context.TODO(), &exp)
+		if err != nil {
+			return nil, err
+		}
+		var fileTree []*experimentv1.FileNode
+		arc, err := archive.FromTarGz(exp.ModelDefinition)
+		if err != nil {
+			return nil, err
+		}
+		err = os.MkdirAll(f.genPath(expID, ""), fs.ModePerm)
+		if err != nil {
+			return nil, err
+		}
+		for _, ar := range arc {
+			if ar.IsDir() {
+				err = os.MkdirAll(f.genPath(expID, ar.Path), fs.ModePerm)
+			} else {
+				err = os.WriteFile(f.genPath(expID, ar.Path), ar.Content, fs.ModePerm)
+			}
```

Comment:
What is the desired behavior for symbolic links?

Reply:
I tried to test this with symbolic links like `ln -s const.yaml link.yaml` then `det e create const.yaml .`, it doesn't seem to make any difference, `link.yaml` seems to be just a regular plain text file

Reply:
I see.  You are right, we aren't actually preserving symbolic links at upload time as I thought we were.

No need to worry about this for now I guess.